writeaheadlog_silo_test.go

```diff
@@ -0,0 +1,332 @@
+package writeaheadlog
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/fastrand"
+	"golang.org/x/crypto/blake2b"
+)
+
+type (
+	silo struct {
+		// offset is the file offset in the fake database
+		offset int64
+
+		// numbers
+		numbers []uint32
+
+		// f is the file on which the silo is stored
+		f *os.File
+
+		// i is the index of the number that needs to be incremented next
+		i uint32
+
+		// cs is the checksum of the silo
+		cs checksum
+	}
+
+	siloUpdate struct {
+		// offset is the in the file at which the number should be written
+		offset int64
+
+		// number is the number which is written at index
+		number uint32
+
+		// silo is the offset of the silo so we can apply the update to the right silo
+		silo int64
+	}
+)
+
+// newSilo creates a new silo of a certain length
+func newSilo(offset int64, length int, f *os.File) *silo {
+	if length == 0 {
+		panic("numbers shouldn't be empty")
+	}
+
+	return &silo{
+		offset:  offset,
+		numbers: make([]uint32, length, length),
+		f:       f,
+	}
+}
+
+// marshal marshals a siloUpdate
+func (su siloUpdate) marshal() []byte {
+	data := make([]byte, 24)
+	binary.PutVarint(data[0:8], su.offset)
+	binary.LittleEndian.PutUint32(data[8:12], su.number)
+	binary.PutVarint(data[12:], su.silo)
+	return data
+}
+
+// unmarshal unmarshals a siloUpdate from data
+func (su *siloUpdate) unmarshal(data []byte) {
+	if len(data) != 24 {
+		panic("data has wrong size")
+	}
+	su.offset, _ = binary.Varint(data[0:8])
+	su.number = binary.LittleEndian.Uint32(data[8:12])
+	su.silo, _ = binary.Varint(data[12:])
+	return
+}
+
+// newUpdate create a WAL update using an index and a number
+func newUpdate(su siloUpdate) Update {
+	update := Update{
+		Name:         "This is my update. There are others like it but this one is mine",
+		Version:      "v0.9.8.7.6.5.4.3.2.1.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z",
+		Instructions: su.marshal(),
+	}
+	return update
+}
+
+// newSiloUpdate creates a new Silo update
+func (s *silo) newSiloUpdate(index uint32, number uint32) siloUpdate {
+	return siloUpdate{
+		number: number,
+		offset: s.offset + int64(4*(index)),
+		silo:   s.offset,
+	}
+}
+
+// updateChecksum calculates the silos checksum and updates the silo's
+// checksum field
+func (s *silo) updateChecksum() {
+	buf := make([]byte, 4*len(s.numbers))
+	for i := 0; i < len(s.numbers); i++ {
+		binary.LittleEndian.PutUint32(buf[i*4:i*4+4], s.numbers[i])
+	}
+	cs := blake2b.Sum256(buf)
+	copy(s.cs[:], cs[0:checksumSize])
```

Comment:
`copy` will copy at most `len(s.cs)` bytes, so you can just do `copy(s.cs[:], cs[:])`. Even better would be `copy(s.cs[:], blake2b.Sum256(buf))`, but I think `Sum256` returns an array and `copy` doesn't like arrays.

Reply:
Yep I tried to do it in one line but copy got upset.